File: domain/src/main/java/com/arturogutierrez/openticator/domain/otp/signer/MacSignerFactory.java
package com.arturogutierrez.openticator.domain.otp.signer;

import com.arturogutierrez.openticator.domain.helper.Base32String;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import javax.inject.Inject;

public class MacSignerFactory {

  @Inject
  public MacSignerFactory() {

  }

  public MacSigner createMacSigner(String secret) throws IllegalArgumentException {
    try {
      byte[] secretBase32 = Base32String.decode(secret);

      Mac mac = Mac.getInstance("HMACSHA1");
      mac.init(new SecretKeySpec(secretBase32, ""));

      return new MacSigner(mac);
    } catch (Exception e) {
      throw new IllegalArgumentException(e.getMessage());
    }
  }
}

File: domain/src/main/kotlin/com/arturogutierrez/openticator/domain/otp/signer/MacSignerFactory.kt
package com.arturogutierrez.openticator.domain.otp.signer

import com.arturogutierrez.openticator.helper.Base32String
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

class MacSignerFactory {

  fun createMacSigner(secret: String): MacSigner {
    try {
      val secretBase32 = Base32String.decode(secret)
      val mac = Mac.getInstance("HMACSHA1")
      mac.init(SecretKeySpec(secretBase32, ""))

      return MacSigner(mac)
    } catch (e: Exception) {
      throw IllegalArgumentException(e.message)
    }
  }
}